frontend/src/redux/middleware.js

```diff
@@ -0,0 +1,55 @@
+import { addWidget } from '@metacell/geppetto-meta-client/common/layout/actions';
+import { WidgetStatus } from '@metacell/geppetto-meta-client/common/layout/model';
+import { v4 as uuidv4 } from 'uuid';
+import { ADD_VIEWER } from './actions/viewers';
+import { ADD_DEVSTAGE } from './actions/devStages';
+import { raiseError, waitData } from './actions/misc';
+import { DevStageService } from '../services/DevStageService';
+
+const devStagesService = new DevStageService();
+
+const middleware = (store) => (next) => (action) => {
+  function widgetFromViewerSpec(viewerSpec) {
+    return {
+      id: viewerSpec.viewerId,
+      name: `${viewerSpec.type}_${viewerSpec.viewerId}`,
+      component: viewerSpec.type,
+      panelName: 'centralPanel',
+      enableClose: true,
+      enableRename: true,
+      enableDrag: true,
+      status: WidgetStatus.ACTIVE,
+      config: {
+        viewerId: viewerSpec.viewerId,
+      },
+    };
+  }
+
+  switch (action.type) {
+    case ADD_DEVSTAGE: {
+      next(waitData('Getting development stages...', action.type));
+      devStagesService.getDevStages().then((stages) => {
+        // eslint-disable-next-line no-param-reassign
```

Comment:
@lrebscher I added this because I couldn't think about an alternative. Is this fine or am I breaking some principle?

Reply:
yeah you shouldn't change the action object in the middleware, instead you would dispatch here another action sth like `RECEIVED_DEVSTAGE` when the request finished, this action would be handled in the reducer. The first action is in this case just a trigger reflecting a user action.

See [here](https://redux.js.org/tutorials/fundamentals/part-6-async-logic#redux-async-data-flow) for a nice visualization of the flow.